File: packages/lynx-clipboard/ios/ClipboardModule.swift
import UIKit
import Lynx

/// Native clipboard access module.
/// JS usage: NativeModules.Clipboard.setString("hello")
class ClipboardModule: NSObject, LynxModule {

    @objc static var name: String { "Clipboard" }

    @objc static var methodLookup: [String: String] {
        [
            "setString": NSStringFromSelector(#selector(setString(_:callback:))),
            "getString": NSStringFromSelector(#selector(getString(_:))),
            "hasString": NSStringFromSelector(#selector(hasString(_:))),
        ]
    }

    required override init() { super.init() }
    required init(param: Any) { super.init() }

    /// Writes to the pasteboard and reports the outcome.
    ///
    /// The callback exists for symmetry with Android, where `setPrimaryClip`
    /// genuinely can throw (a clip over the binder transaction limit). Nothing
    /// here can fail, so this always reports success — but a caller awaiting
    /// `setString` must get an answer on both platforms, or the promise hangs
    /// on iOS.
    @objc func setString(_ text: String?, callback: LynxCallbackBlock?) {
        UIPasteboard.general.string = text ?? ""
        callback?([:])
    }

    @objc func getString(_ callback: LynxCallbackBlock?) {
        let text = UIPasteboard.general.string ?? ""
        callback?(text)
    }

    @objc func hasString(_ callback: LynxCallbackBlock?) {
        let has = UIPasteboard.general.hasStrings
        callback?(has)
    }
}

File: packages/lynx-clipboard/android/com/sigx/clipboard/ClipboardModule.kt
package com.sigx.clipboard

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyMap

/**
 * Native clipboard access module.
 * JS usage: NativeModules.Clipboard.setString("hello")
 *           NativeModules.Clipboard.getString(callback)
 */
class ClipboardModule(context: Context) : LynxModule(context) {

    private val clipboard: ClipboardManager by lazy {
        mContext.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    }

    /**
     * Writes to the clipboard and reports the outcome.
     *
     * `setPrimaryClip` really can fail — a clip larger than the binder
     * transaction limit throws, and OEM clipboard managers have been known to
     * as well. That failure used to be invisible to JS because this method
     * returned nothing.
     */
    @LynxMethod
    fun setString(text: String?, callback: Callback?) {
        try {
            val clip = ClipData.newPlainText("sigx-lynx-go", text ?: "")
            clipboard.setPrimaryClip(clip)
            callback?.invoke(JavaOnlyMap())
        } catch (e: Throwable) {
            val map = JavaOnlyMap()
            map.putString("error", e.message ?: "setPrimaryClip failed")
            callback?.invoke(map)
        }
    }

    @LynxMethod
    fun getString(callback: Callback?) {
        val text = clipboard.primaryClip
            ?.getItemAt(0)
            ?.text
            ?.toString() ?: ""
        callback?.invoke(text)
    }

    @LynxMethod
    fun hasString(callback: Callback?) {
        val has = clipboard.hasPrimaryClip() &&
                clipboard.primaryClipDescription?.hasMimeType("text/*") == true
        callback?.invoke(has)
    }
}

